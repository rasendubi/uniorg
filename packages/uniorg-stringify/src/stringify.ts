import type {
  AffiliatedKeywords,
  OrgNode,
  WithAffiliatedKeywords,
} from 'uniorg';
import type { Node } from 'unist';

type Handler<T> = (org: T, options: StringifyOptions) => string | null;

export type Handlers = {
  [K in OrgNode['type']]?: Handler<OrgNode & { type: K }>;
};

export type StringifyOptions = {
  handlers: Handlers;
};

const defaultOptions: StringifyOptions = {
  handlers: {
    'export-snippet': (org) => `@@${org.backEnd}:${org.value}@@`,
  },
};

function normalizeOptions(
  options: Partial<StringifyOptions>
): StringifyOptions {
  return {
    ...defaultOptions,
    ...options,
    // Merging handlers separately to allow overriding it partially.
    handlers: {
      ...defaultOptions.handlers,
      ...options.handlers,
    },
  };
}

export function stringify(
  org: string | Node | Node[],
  options: Partial<StringifyOptions> = {}
): string {
  const fullOptions = normalizeOptions(options);
  const result = Array.isArray(org)
    ? org.map((o) => stringify(o, fullOptions)).join('')
    : stringifyOne(org, fullOptions);
  return result;
}

function stringifyOne(node: Node | string, options: StringifyOptions): string {
  if (typeof node === 'string') {
    return node;
  }

  const org = node as OrgNode & Partial<WithAffiliatedKeywords>;

  const result: string[] = [];

  if (org.affiliated) {
    result.push(
      stringifyAffiliated(org.affiliated as AffiliatedKeywords, options)
    );
  }

  result.push(stringifyNode(org, options));

  return result.join('');
}

function stringifyNode(org: OrgNode, options: StringifyOptions): string {
  const handler = options.handlers?.[org.type];
  if (handler) {
    const rendered = (handler as any)(org, options);
    if (rendered !== null && rendered !== undefined) {
      return rendered;
    }
  }

  switch (org.type) {
    case 'org-data':
      return withNewline(stringify(org.children, options));
    case 'section':
      return withNewline(stringify(org.children, options));
    case 'headline': {
      const components = [
        '*'.repeat(org.level),
        org.todoKeyword,
        org.priority ? `[#${org.priority}]` : null,
        org.commented ? 'COMMENT' : null,
        stringify(org.children, options),
        org.tags.length ? `:${org.tags.join(':')}:` : null,
      ].filter((x) => x !== null);
      return withNewline(components.join(' '));
    }
    case 'statistics-cookie':
      return [org.value, ' '.repeat(org.postBlank)].join('');
    case 'planning':
      return withNewline(
        [
          org.closed ? `CLOSED: ${stringify(org.closed, options)}` : null,
          org.scheduled
            ? `SCHEDULED: ${stringify(org.scheduled, options)}`
            : null,
          org.deadline ? `DEADLINE: ${stringify(org.deadline, options)}` : null,
        ]
          .filter((x) => x !== null)
          .join(' ')
      );
    case 'property-drawer':
      return withNewline(
        [
          ':PROPERTIES:\n',
          ...org.children.map((c) => stringify(c, options)),
          ':END:',
        ].join('')
      );
    case 'node-property':
      return withNewline([':', org.key, ': ', org.value].join(''));
    case 'drawer':
      return [
        ':',
        org.name,
        ':\n',
        withNewline(stringify(org.children, options)),
        ':END:\n',
      ].join('');
    case 'keyword':
      return withNewline(`#+${org.key}: ${org.value}`);
    case 'plain-list':
      return withNewline(stringify(org.children, options));
    case 'list-item':
      return withNewline(
        [
          ' '.repeat(org.indent),
          org.bullet,
          org.counter,
          org.checkbox === 'on'
            ? '[X] '
            : org.checkbox === 'off'
              ? '[ ] '
              : org.checkbox === 'trans'
                ? '[-] '
                : null,
          indent(
            stringify(org.children, options),
            org.indent + org.bullet.length
          ).trimStart(),
        ]
          .filter((x) => x !== null)
          .join('')
      );
    case 'list-item-tag':
      return `${stringify(org.children[0], options)} :: ${stringify(
        org.children.slice(1),
        options
      )}`;
    case 'table':
      const value =
        org.tableType === 'table.el'
          ? org.value
          : stringify(org.children, options);
      return (
        withNewline(value) + (org.tblfm ? '#+TBLFM: ' + org.tblfm + '\n' : '')
      );
    case 'table-row':
      if (org.rowType === 'standard') {
        return (
          '| ' +
          org.children.map((c) => stringify(c, options)).join(' | ') +
          ' |\n'
        );
      } else {
        return '|-|\n';
      }
    case 'table-cell':
      return stringify(org.children, options);
    case 'comment':
      return withNewline(
        org.value
          .split(/\n/)
          .map((s) => '# ' + s)
          .join('\n')
      );
    case 'fixed-width':
      return withNewline(
        org.value
          .split(/\n/)
          .map((s) => ': ' + s)
          .join('\n')
      );
    case 'clock':
      return withNewline(
        [
          'CLOCK: ',
          org.value ? stringify(org.value, options) : null,
          org.duration ? ' =>  ' + org.duration : null,
        ]
          .filter((x) => x !== null)
          .join('')
      );
    case 'latex-environment':
      return withNewline(org.value);
    case 'horizontal-rule':
      return withNewline('-----');
    case 'footnote-definition':
      return withNewline(
        `[fn:${org.label}] ${stringify(org.children, options)}`
      );
    case 'footnote-reference':
      return [
        '[fn:',
        org.label,
        org.footnoteType === 'inline'
          ? ':' + stringify(org.children, options)
          : null,
        ']',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'diary-sexp':
      return withNewline(org.value);
    case 'src-block':
      return [
        '#+begin_src',
        org.language ? ' ' + org.language : null,
        '\n',
        withNewline(escapeCodeInString(org.value)),
        '#+end_src\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'quote-block':
      return [
        '#+begin_quote\n',
        withNewline(stringify(org.children, options)),
        '#+end_quote\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'verse-block':
      return [
        '#+begin_verse\n',
        withNewline(stringify(org.children, options)),
        '#+end_verse\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'center-block':
      return [
        '#+begin_center\n',
        withNewline(stringify(org.children, options)),
        '#+end_center\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'comment-block':
      return ['#+begin_comment\n', withNewline(org.value), '#+end_comment\n']
        .filter((x) => x !== null)
        .join('');
    case 'example-block':
      return ['#+begin_example\n', withNewline(org.value), '#+end_example\n']
        .filter((x) => x !== null)
        .join('');
    case 'export-block':
      return [
        '#+begin_export',
        org.backend ? ' ' + org.backend : null,
        '\n',
        withNewline(org.value),
        '#+end_export\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'special-block':
      return [
        '#+begin_',
        org.blockType,
        '\n',
        withNewline(stringify(org.children, options)),
        '#+end_',
        org.blockType,
        '\n',
      ]
        .filter((x) => x !== null)
        .join('');
    case 'paragraph':
      return (
        withNewline(stringify(org.children, options)) +
        // an extra newline to separate from the possible following paragraph
        '\n'
      );

    case 'citation':
      return [
        '[cite',
        org.style ? '/' + org.style : '',
        ':',
        ...org.children.map((c) => stringifyNode(c, options)).join(';'),
        ']',
      ].join('');
    case 'citation-common-prefix':
    case 'citation-common-suffix':
    case 'citation-reference':
    case 'citation-prefix':
    case 'citation-suffix':
      return org.children.map((c) => stringifyNode(c, options)).join('');
    case 'citation-key':
      return '@' + org.key;

    case 'link':
      return org.format === 'plain'
        ? org.rawLink
        : org.format === 'bracket'
          ? `[[${org.rawLink}]${
              org.children.length
                ? '[' + stringify(org.children, options) + ']'
                : ''
            }]`
          : `<${org.rawLink}>`;
    case 'superscript':
      return `^{${stringify(org.children, options)}}`;
    case 'subscript':
      return `_{${stringify(org.children, options)}}`;
    case 'bold':
      return `*${stringify(org.children, options)}*`;
    case 'italic':
      return `/${stringify(org.children, options)}/`;
    case 'strike-through':
      return `+${stringify(org.children, options)}+`;
    case 'underline':
      return `_${stringify(org.children, options)}_`;
    case 'code':
      return `~${org.value}~`;
    case 'verbatim':
      return `=${org.value}=`;
    case 'latex-fragment':
      return org.value;
    case 'timestamp':
      return org.rawValue;
    case 'entity':
      return '\\' + org.name;
    case 'text':
      return org.value;
    default:
      return '';
  }
}

function stringifyAffiliated(
  keywords: AffiliatedKeywords,
  options: StringifyOptions
): string {
  return Object.entries(keywords)
    .map(([key, values]) => {
      const dualKeyword = Array.isArray(values) ? values[1] : null;
      const value = Array.isArray(values) ? values[0] : values;
      return [
        '#+',
        key,
        // TODO: "as any" here is incorrect
        dualKeyword ? `[${stringify(dualKeyword as any, options)}]` : null,
        ': ',
        // TODO: "as any" here is incorrect
        stringify(value as any, options).trimEnd(),
        '\n',
      ].join('');
    })
    .join('');
}

function withNewline(s: string): string {
  return s.trimEnd() + '\n';
}

function indent(s: string, level: number): string {
  // strip at most `level` spaces
  const stripRe = new RegExp(`^ {0,${level}}`);
  const indent = ' '.repeat(level);
  return s
    .split(/\n/)
    .map((line) => indent + line.replace(stripRe, ''))
    .join('\n');
}

function escapeCodeInString(value: string): string {
  return value.replace(/^([ \t]*)(,*)(\*|#\+)/gm, '$1,$2$3');
}
