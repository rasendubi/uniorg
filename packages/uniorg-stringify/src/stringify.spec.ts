import { parse } from 'uniorg-parse/lib/parser';

import { StringifyOptions, stringify } from './stringify';

const process = (input: string, options: Partial<StringifyOptions>) => {
  return stringify(parse(input), options);
};

const test = (
  name: string,
  input: string,
  options: Partial<StringifyOptions> = {}
) => {
  it(name, () => {
    const result = process(input, options);
    expect(result).toMatchSnapshot();
  });
  it(name + ' (second pass)', () => {
    const result1 = process(input, options);
    const result2 = process(result1, options);
    expect(result2).toEqual(result1);
  });
};
test.skip = (
  name: string,
  input: string,
  options: Partial<StringifyOptions> = {}
) => {
  it.skip(name, () => {
    const result = process(input, options);
    expect(result).toMatchSnapshot();
  });
  it.skip(name + ' (second pass)', () => {
    const result1 = process(input, options);
    const result2 = process(result1, options);
    expect(result2).toEqual(result1);
  });
};
test.only = (
  name: string,
  input: string,
  options: Partial<StringifyOptions> = {}
) => {
  it.only(name, () => {
    const result = process(input, options);
    expect(result).toMatchSnapshot();
  });
  it.only(name + ' (second pass)', () => {
    const result1 = process(input, options);
    const result2 = process(result1, options);
    expect(result2).toEqual(result1);
  });
};

describe('stringify', () => {
  test('empty input', '');

  test('paragraph', 'hello, world!');
  test(
    'two paragraphs',
    `first paragraph

second paragraph`
  );

  describe('headline', () => {
    test('simple headline', '* Hello');
    test(
      'complex headline',
      `* TODO [#A] COMMENT headline /italic/ title :some:tags:`
    );
    test(
      'nested headline',
      `* level 1
** level 2

* level 1
*** level 3`
    );
    test(
      'nested simple headlines separated by new-lines',
      `* level 1

** level 2

*** level 3

**** level 4`,
      {
        handlers: {
          headline: (org, options) => {
            const components = [
              '*'.repeat(org.level),
              stringify(org.children, options),
            ].filter((x) => x !== null);
            return `${components.join(' ')}\n\n`;
          },
        },
      }
    );

    test('statistics cookie', `* [115%]  headline`);
  });

  test(
    'planning',
    `* headline
CLOSED: [2019-03-13 Wed 23:48] SCHEDULED: [2019-03-13 Wed] DEADLINE: [2019-03-14 Thu]`
  );
  test(
    'mixed good and fake planning',
    `* headline
CLOSED: SCHEDULED: [2021-05-31 Mon]
`
  );

  test(
    'property drawer',
    `* headline
:PROPERTIES:
:CREATED: [2019-03-13 Wed 23:57]
:END:`
  );
  test(
    'property drawer forced to lower-case',
    `* headline
:PROPERTIES:
:CREATED: [2019-03-13 Wed 23:57]
:END:`,
    {
      handlers: {
        'property-drawer': (org, options) => {
          return (
            [
              ':properties:\n',
              ...org.children.map((c) => stringify(c, options)),
              ':end:',
            ].join('') + `\n`
          );
        },
        'node-property': (org) => {
          return [':', org.key.toLowerCase(), ': ', org.value].join('') + '\n';
        },
      },
    }
  );
  test(
    'file property drawer',
    `:PROPERTIES:
:ID:       01c7615e-d792-4d06-995b-19a2c046c055
:END:`
  );

  test(
    'custom drawer',
    `:MYDRAWER:
hello /there/
:END:`
  );

  describe('list', () => {
    test('single-item list', `- hi`);
    test(
      'ordered list',
      `1. one
2. two`
    );
    test(
      'nested lists',
      `- there
  - nested
  - list`
    );
    test(
      'description list',
      `- term1 :: description 1
- term 2 :: description 2`
    );
    test(
      'maintains indentation in list',
      `
- list item 1
  more nesting
  - list 1.1
    :DRAWER:
    hello
    :END:
    more text
  - list 1.2

    still in 1.2`
    );

    test('checkbox list', `- [ ] hello`);
  });

  describe('links', () => {
    test('link', `http://example.com`);

    test('regular link', `[[link][text]]`);

    test('link with no text', `[[link]]`);

    test('file link with spaces', `[[./file with spaces.org]]`);

    // note that these are actual percents in the file name, not a url-encoded "file with spaces.org"
    test('file link with percents', `[[./file%2Bwith%2Bspaces.org]]`);

    test('angle link', `<http://example.com>`);

    test('https link', `[[https://example.com/hello]]`);

    test(
      'multiline description',
      `[[www.something.com][line1
line2]]`
    );
  });

  describe('blocks', () => {
    test(
      'src block',
      `#+begin_src
hello
#+end_src`
    );

    test(
      'escaper in src block',
      `#+begin_src c
,*a = 0;
#+end_src`
    );

    test(
      'headline escaper in src block',
      `#+begin_src org
,* not a headline;
#+end_src`
    );

    test(
      'comma escaper in src block',
      `#+begin_src org
,,,* two commas escaped
#+end_src`
    );

    test(
      '#+ escaper in src block',
      `#+begin_src org
,#+ escaped
#+end_src`
    );

    test(
      'fake escaper',
      `#+begin_src org
,# nont escaped
#+end_src`
    );

    test(
      'src in list',
      `
- example:
  #+begin_src
  blah
  #+end_src`
    );

    test(
      'quote block',
      `#+begin_quote
hello
#+end_quote`
    );

    test(
      'verse block',
      `#+BEGIN_VERSE
 Great clouds overhead
 Tiny black birds rise and fall
 Snow covers Emacs

    ---AlexSchroeder
#+END_VERSE`
    );

    test(
      'center block',
      `#+begin_center
hello
#+end_center`
    );

    test(
      'comment block',
      `#+begin_comment
hello
#+end_comment`
    );

    test(
      'example block',
      `#+begin_example
hi
#+end_example`
    );

    test(
      'export block',
      `#+begin_export
hi
#+end_export`
    );

    test(
      'export block with backend',
      `#+begin_export html
hi
#+end_export`
    );

    test(
      'special block',
      `#+begin_blah
hello
#+end_blah`
    );
  });

  describe('timestamps', () => {
    test('inactive', `[2021-01-07 Thu]`);
    test('inactive-range', `[2021-01-07 Thu]--[2021-01-08 Fri]`);
    test('active', `<2021-01-07 Thu>`);
    test('active-range', `<2021-01-07 Thu>--<2021-01-09 Sat>`);
    test('with time', `[2021-01-07 Thu 19:36]`);
    test('time range', `[2021-01-07 Thu 19:36-20:38]`);
  });

  describe('keyword', () => {
    test('keyword', `#+title: hi`);
    test(
      'multiple keywords',
      `
#+title: hi
#+author: me`
    );
    test(
      'dual keyword',
      `#+RESULTS[hi]: hello
there`
    );
    test(
      'multiple affiliated keywords',
      `
#+name: name
#+caption: caption
paragraph
`
    );
  });

  test('horizontal rule', `------`);

  describe('footnotes', () => {
    test(
      'footnote definition',
      `[fn:hello] this is footnote definition
`
    );
    test(
      'starting on next line',
      `
[fn:hello]
footnote
`
    );
    test(
      'sequential',
      `
[fn:hello] footnote1
[fn:2] footnote2
`
    );

    // note that footnote content is ": footnote", not "footnote".
    test('footnote definition with : following', `[fn:1]: footnote`);

    test('standard reference', `hello[fn:1]`);
    test(
      'inline reference',
      `hello[fn:: this is inline /footnote/ definition]`
    );
    test(
      'named inline reference',
      `hello[fn:name: this is inline /footnote/ definition]`
    );
  });

  test(
    'diary sexp',
    `%%(diary-anniversary 10 31 1948) Arthur's birthday (%d years old)`
  );

  describe('emphasis marks', () => {
    test('bold', '*hello*');
    test('bold with custom char', '*hello*', {
      handlers: {
        bold: (org, options) => {
          return `$${stringify(org.children, options)}$`;
        },
      },
    });
    test('emphasis', `/Consider/ ~t*h*e~ *following* =example= +strike+`);
    test('underline', `_hello_`);
  });

  describe('superscript', () => {
    test('simple', 'hello^world');
    test('with braces', `H^{+}`);
    test('nested braces', `H^{{x}}`);
  });
  describe('subscript', () => {
    test('simple', 'hello_world');
    test('with braces', `H_{+}`);
    test('nested braces', `H_{{x}}`);
  });

  test(
    'latex-fragment',
    `If $a^2=b$ and \\( b=2 \\), then the solution must be
either $$ a=+\\sqrt{2} $$ or \\[ a=-\\sqrt{2} \\].`
  );

  test('entity', `\\Aggrave`);

  test(
    'table',
    `
| head1  | head2 |
|--------+-------|
| value1 | value2 |
`
  );
  test(
    'table.el table',
    `
+--------+--------+
| head1  | head2  |
+--------+--------+
| value1 | value2 |
+--------+--------+`
  );

  test(
    'table with format',
    `
| 1 |
| 2 |
#+TBLFM: $2=$1^2::$3=$1^3
`
  );

  test('single-line comment', `# this is comment`);
  test(
    'multi-line comment',
    `# first line
  # second
# third
                    # fourth`
  );

  test(
    'multi-line fixed-width',
    `: hello
:   world`
  );

  describe('clock', () => {
    test('clock in progress', `CLOCK: [2021-01-10 Sun 14:36]`);
    test(
      'finished clock',
      `CLOCK: [2020-12-22 Tue 09:07]--[2020-12-22 Tue 11:10] =>  2:03`
    );
  });

  test(
    'multi-line latex environment',
    `\\begin{hello}
some text
\\end{hello}`
  );

  test(
    'citation',
    `[cite/style:common prefix; prefix @key suffix; @key2; common suffix]`
  );

  describe('handlers', () => {
    test(
      'allow handlers to return empty string to suppress output',
      `hello, +Dave+ world`,
      { handlers: { 'strike-through': () => '' } }
    );

    // This usage is currently not guaranteed and may change in the future.
    test.skip(
      'allow handlers to return null to fallback to default processing',
      `hello, +Dave+ world`,
      { handlers: { 'strike-through': () => null } }
    );
  });

  test('handle export-snippet', `@@backend:custom value@@`);

  test(
    'line-break',
    `hello\\\\
world!`
  );
  test('line-break with trailing whitespace', `hello\\\\  \nworld!`);
});
