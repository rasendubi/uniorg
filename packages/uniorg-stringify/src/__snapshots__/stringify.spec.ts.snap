// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify blocks #+ escaper in src block 1`] = `
"#+begin_src org
,#+ escaped
#+end_src
"
`;

exports[`stringify blocks center block 1`] = `
"#+begin_center
hello
#+end_center
"
`;

exports[`stringify blocks comma escaper in src block 1`] = `
"#+begin_src org
,,,* two commas escaped
#+end_src
"
`;

exports[`stringify blocks comment block 1`] = `
"#+begin_comment
hello
#+end_comment
"
`;

exports[`stringify blocks escaper in src block 1`] = `
"#+begin_src c
,*a = 0;
#+end_src
"
`;

exports[`stringify blocks example block 1`] = `
"#+begin_example
hi
#+end_example
"
`;

exports[`stringify blocks export block 1`] = `
"#+begin_export
hi
#+end_export
"
`;

exports[`stringify blocks export block with backend 1`] = `
"#+begin_export html
hi
#+end_export
"
`;

exports[`stringify blocks fake escaper 1`] = `
"#+begin_src org
,# nont escaped
#+end_src
"
`;

exports[`stringify blocks headline escaper in src block 1`] = `
"#+begin_src org
,* not a headline;
#+end_src
"
`;

exports[`stringify blocks quote block 1`] = `
"#+begin_quote
hello
#+end_quote
"
`;

exports[`stringify blocks special block 1`] = `
"#+begin_blah
hello
#+end_blah
"
`;

exports[`stringify blocks src block 1`] = `
"#+begin_src
hello
#+end_src
"
`;

exports[`stringify blocks src in list 1`] = `
"- example:
  
  #+begin_src
  blah
  #+end_src
"
`;

exports[`stringify blocks verse block 1`] = `
"#+begin_verse
 Great clouds overhead
 Tiny black birds rise and fall
 Snow covers Emacs

    ---AlexSchroeder
#+end_verse
"
`;

exports[`stringify citation 1`] = `
"[cite/style:common prefix; prefix @key suffix;@key2; common suffix]
"
`;

exports[`stringify clock clock in progress 1`] = `
"CLOCK: [2021-01-10 Sun 14:36]
"
`;

exports[`stringify clock finished clock 1`] = `
"CLOCK: [2020-12-22 Tue 09:07]--[2020-12-22 Tue 11:10] =>  2:03
"
`;

exports[`stringify custom drawer 1`] = `
":MYDRAWER:
hello /there/
:END:
"
`;

exports[`stringify diary sexp 1`] = `
"%%(diary-anniversary 10 31 1948) Arthur's birthday (%d years old)
"
`;

exports[`stringify emphasis marks bold 1`] = `
"*hello*
"
`;

exports[`stringify emphasis marks bold with custom char 1`] = `
"$hello$
"
`;

exports[`stringify emphasis marks emphasis 1`] = `
"/Consider/ ~t*h*e~ *following* =example= +strike+
"
`;

exports[`stringify emphasis marks underline 1`] = `
"_hello_
"
`;

exports[`stringify empty input 1`] = `
"
"
`;

exports[`stringify entity 1`] = `
"\\\\Aggrave
"
`;

exports[`stringify file property drawer 1`] = `
":PROPERTIES:
:ID: 01c7615e-d792-4d06-995b-19a2c046c055
:END:
"
`;

exports[`stringify footnotes footnote definition 1`] = `
"[fn:hello] this is footnote definition
"
`;

exports[`stringify footnotes footnote definition with : following 1`] = `
"[fn:1] : footnote
"
`;

exports[`stringify footnotes inline reference 1`] = `
"hello[fn:: this is inline /footnote/ definition]
"
`;

exports[`stringify footnotes named inline reference 1`] = `
"hello[fn:name: this is inline /footnote/ definition]
"
`;

exports[`stringify footnotes sequential 1`] = `
"[fn:hello] footnote1
[fn:2] footnote2
"
`;

exports[`stringify footnotes standard reference 1`] = `
"hello[fn:1]
"
`;

exports[`stringify footnotes starting on next line 1`] = `
"[fn:hello] footnote
"
`;

exports[`stringify handle export-snippet 1`] = `
"@@backend:custom value@@
"
`;

exports[`stringify handlers allow handlers to return empty string to suppress output 1`] = `
"hello,  world
"
`;

exports[`stringify handlers allow handlers to return null to fallback to default processing 1`] = `
"hello, +Dave+ world
"
`;

exports[`stringify headline complex headline 1`] = `
"* TODO [#A] COMMENT headline /italic/ title :some:tags:
"
`;

exports[`stringify headline nested headline 1`] = `
"* level 1
** level 2
* level 1
*** level 3
"
`;

exports[`stringify headline nested simple headlines separated by new-lines 1`] = `
"* level 1

** level 2

*** level 3

**** level 4
"
`;

exports[`stringify headline simple headline 1`] = `
"* Hello
"
`;

exports[`stringify headline statistics cookie 1`] = `
"* [115%]  headline
"
`;

exports[`stringify horizontal rule 1`] = `
"-----
"
`;

exports[`stringify keyword dual keyword 1`] = `
"#+RESULTS[hi]: hello
there
"
`;

exports[`stringify keyword keyword 1`] = `
"#+TITLE: hi
"
`;

exports[`stringify keyword multiple affiliated keywords 1`] = `
"#+NAME: name
#+CAPTION: caption
paragraph
"
`;

exports[`stringify keyword multiple keywords 1`] = `
"#+TITLE: hi
#+AUTHOR: me
"
`;

exports[`stringify latex-fragment 1`] = `
"If $a^2=b$ and \\\\( b=2 \\\\), then the solution must be
either $$ a=+\\\\sqrt{2} $$ or \\\\[ a=-\\\\sqrt{2} \\\\].
"
`;

exports[`stringify links angle link 1`] = `
"<http://example.com>
"
`;

exports[`stringify links file link with percents 1`] = `
"[[./file%2Bwith%2Bspaces.org]]
"
`;

exports[`stringify links file link with spaces 1`] = `
"[[./file with spaces.org]]
"
`;

exports[`stringify links https link 1`] = `
"[[https://example.com/hello]]
"
`;

exports[`stringify links link 1`] = `
"http://example.com
"
`;

exports[`stringify links link with no text 1`] = `
"[[link]]
"
`;

exports[`stringify links multiline description 1`] = `
"[[www.something.com][line1
line2]]
"
`;

exports[`stringify links regular link 1`] = `
"[[link][text]]
"
`;

exports[`stringify list checkbox list 1`] = `
"- [ ] hello
"
`;

exports[`stringify list description list 1`] = `
"- term1 :: description 1
- term 2 :: description 2
"
`;

exports[`stringify list maintains indentation in list 1`] = `
"- list item 1
  more nesting
  
  - list 1.1
    
    :DRAWER:
    hello
    :END:
    more text
  - list 1.2
    
    still in 1.2
"
`;

exports[`stringify list nested lists 1`] = `
"- there
  
  - nested
  - list
"
`;

exports[`stringify list ordered list 1`] = `
"1. one
2. two
"
`;

exports[`stringify list single-item list 1`] = `
"- hi
"
`;

exports[`stringify mixed good and fake planning 1`] = `
"* headline
SCHEDULED: [2021-05-31 Mon]
"
`;

exports[`stringify multi-line comment 1`] = `
"# first line
# second
# third
# fourth
"
`;

exports[`stringify multi-line fixed-width 1`] = `
": hello
:   world
"
`;

exports[`stringify multi-line latex environment 1`] = `
"\\\\begin{hello}
some text
\\\\end{hello}
"
`;

exports[`stringify paragraph 1`] = `
"hello, world!
"
`;

exports[`stringify planning 1`] = `
"* headline
CLOSED: [2019-03-13 Wed 23:48] SCHEDULED: [2019-03-13 Wed] DEADLINE: [2019-03-14 Thu]
"
`;

exports[`stringify property drawer 1`] = `
"* headline
:PROPERTIES:
:CREATED: [2019-03-13 Wed 23:57]
:END:
"
`;

exports[`stringify property drawer forced to lower-case 1`] = `
"* headline
:properties:
:created: [2019-03-13 Wed 23:57]
:end:
"
`;

exports[`stringify single-line comment 1`] = `
"# this is comment
"
`;

exports[`stringify subscript nested braces 1`] = `
"H_{{x}}
"
`;

exports[`stringify subscript simple 1`] = `
"hello_{world}
"
`;

exports[`stringify subscript with braces 1`] = `
"H_{+}
"
`;

exports[`stringify superscript nested braces 1`] = `
"H^{{x}}
"
`;

exports[`stringify superscript simple 1`] = `
"hello^{world}
"
`;

exports[`stringify superscript with braces 1`] = `
"H^{+}
"
`;

exports[`stringify table 1`] = `
"| head1 | head2 |
|-|
| value1 | value2 |
"
`;

exports[`stringify table with format 1`] = `
"| 1 |
| 2 |
#+TBLFM: $2=$1^2::$3=$1^3
"
`;

exports[`stringify table.el table 1`] = `
"+--------+--------+
| head1  | head2  |
+--------+--------+
| value1 | value2 |
+--------+--------+
"
`;

exports[`stringify timestamps active 1`] = `
"<2021-01-07 Thu>
"
`;

exports[`stringify timestamps active-range 1`] = `
"<2021-01-07 Thu>--<2021-01-09 Sat>
"
`;

exports[`stringify timestamps inactive 1`] = `
"[2021-01-07 Thu]
"
`;

exports[`stringify timestamps inactive-range 1`] = `
"[2021-01-07 Thu]--[2021-01-08 Fri]
"
`;

exports[`stringify timestamps time range 1`] = `
"[2021-01-07 Thu 19:36-20:38]
"
`;

exports[`stringify timestamps with time 1`] = `
"[2021-01-07 Thu 19:36]
"
`;

exports[`stringify two paragraphs 1`] = `
"first paragraph

second paragraph
"
`;
